Make inspec much faster for most commands.

zenspider opened the pull request:
Removes pre-loading and registration of resources by adding
method_missing to a couple key areas.

This still drops the ball in some areas, but it is a start.

zenspider committed:
Make inspec much faster for most commands.

Removes pre-loading and registration of resources by adding
method_missing to a couple key areas.

This still drops the ball in some areas, but it is a start.

zenspider committed:
I don't like this, but make chef shell require all resources.

I think there is probably a better way to do this, but my first
instinct was that I'd have to dig into hacking pry, and I don't want
to do that right now.

Signed-off-by: Ryan Davis <zenspider@chef.io>

clintoncwolfe commented on lib/inspec/backend.rb:
```diff
@@ -77,6 +77,14 @@ def self.create(config) # rubocop:disable Metrics/AbcSize
           connection
         end
 
+        def method_missing(id, *args, &blk)
```
I feel that at this point we should be careful to heavily explain what is happening when we add new elements of metaprogramming. What context is this being called in? Explain it to most novice new developer on the team, but also to the potential contributor, in comments. Right now it is fresh in your mind and obvious; it will never be easier.

I can follow this one, but I have the tribal knowledge that requiring a resource has the side effect registering it in the registry.

clintoncwolfe replied:
Thanks!

clintoncwolfe commented on lib/inspec/cli.rb:
```diff
@@ -330,6 +323,7 @@ def detect
   option :distinct_exit, type: :boolean, default: true,
     desc: "Exit with code 100 if any tests fail, and 101 if any are skipped but none failed (default).  If disabled, exit 0 on skips and 1 for failures."
   def shell_func
+    require "inspec/resources" # TODO: remove me if possible!
```
I'm not sure we can - we definitely need it for tab completion and help.  If we had multithreading, we could get to a prompt early and then load the heavy stuff while the user types...

miah replied:
I'm curious how many users use shell, autocomplete, or help =)

I think so long as we tell people we are going to break this functionality for a limited time while are making things faster this would be fine.

zenspider replied:
I worked this out too.

clintoncwolfe commented on lib/inspec/control_eval_context.rb:
```diff
@@ -66,9 +66,15 @@ def method_missing(method_name, *arguments, &block)
+              klass = Inspec::Resource.registry[id.to_s]
+              klass.new(inspec, id, *arguments)
+            rescue LoadError
+              # TODO: aws
```
If you're referring to the aws resources in lib/resources, those are deprecated and are scheduled for removal in 5.0 - so I'm not sure it's worth adding code to handle.

miah replied:
If I understand the change correctly, we'll need code to handle AWS until we completely remove it otherwise those resources won't be loaded and users would be forced to use inspec-aws.

zenspider replied:
I needed to handle it to make tests pass, so... done.

clintoncwolfe reviewed (commented):
I think we need to talk about this, as it is actually a breaking change - we currently promise people that Registry is fully populated at the start of a run, and you can cross-reference both core and custom resources that way. So dynamically loading resources (in this implementation) might break that. 

We could fix that by having Registry be more intelligent and perform 2-stage loading - we can glob lib/inspec/resources, which is far cheaper than the actual require.

I also request that we comment the addition - without it, it's just another sudden chunk of metaprogramming in the swamp.  Right now it has a clear purpose, easy to understand.

EDIT - After discussion, so few people use that undocumented feature that it is not worth preserving a full registry. My concerns are addressed.

zenspider committed:
Sped up inspec shell by avoiding pre-registering all reesources.

Now it tries to load a resource when `help xxx` is invoked or loads
all resources when `help resources` is invoked. Otherwise, it stays
lazy.

Signed-off-by: Ryan Davis <zenspider@chef.io>

clintoncwolfe reviewed (changes requested):
I'm happy with the implementation but would request that we add comments explaining that this is auto-loading the resources, and perhaps a comment about requiring a resource resulting in it being registered.

miah reviewed (changes requested):
I agree that we need a comment or documentation update.

It feels like this PR should also delete `inspec/resources.rb` since we're removing all references to it.

zenspider committed:
Refactor into Inspec::DSL.method_missing_resource

This way I can add aws lazy loading to only one place.

I don't actually care that this is in Inspec::DSL. There is probably a
better place for it. Suggestions welcome.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Moved rspec/matchers require to somewhere more appropriate.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Fixed a bug in checking supports (strings vs symbols)

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Documented lib/inspec/resources.rb current usage.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Extended to pick up our internal aws resources as well.

Signed-off-by: Ryan Davis <zenspider@chef.io>

clintoncwolfe reviewed (approved):
Looks great - thanks!

miah commented on lib/inspec/dsl.rb:
```diff
@@ -44,7 +63,11 @@ def method_missing(method_name, *arguments, &block)
     else
-      super
+      begin
+        Inspec::DSL.method_missing_resource(@backend, method_name, *arguments)
```
Is `@backend` right to use here? Can we use the accessor added on line 6?

zenspider replied:
I took a whack at using `backend` and it broke some tests... :/

miah commented on lib/inspec/resources/platform.rb:
```diff
@@ -67,19 +69,22 @@ def supported?(supports)
+      supports.each do |support|
+        support.each do |k, v|
+          status =
+            case k
```
nice!

miah reviewed (changes requested):
Can you answer this question? https://github.com/inspec/inspec/pull/4365/files#r318319942

Approve after that =)

zenspider committed:
Replace backticks in test descriptions with single-quotes.

Backticks means you can't selectively run one of these tests w/o
subshelling out to `brew`, or having to edit first.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Save off supports[:os] and restore it after each test.

This fixes some other tests that require supports[:os] to be correct.

Signed-off-by: Ryan Davis <zenspider@chef.io>
